### PR TITLE
Add support for Meta Devices like Sonos Play.

### PR DESCRIPTION
**Pull request by Tolriq**
On those devices the real devices and services are listed as child devices.

And as a side note a big thanks for https://github.com/ohmae/mmupnp, i can now ditch Cling and it's 15k methods :)

### CODE REVIEW

**Comment by ohmae** [comment]
Thank you!

> On those devices the real devices and services are listed as child devices.

Oh, I did not know that such a device existed.
I can not check behavior because I do not have such device.
I will review your code.

**Comment by Tolriq** [comment]
The Sonos XML so you can verify add tests if wanted.

It's runtime tested and both Server / Renderer correctly works with that patch.

I've added a feature request on the main library if you have time to answer :) Can eventually PR that too.

```
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
<major>1</major>
<minor>0</minor>
</specVersion>
<device>
<deviceType>urn:schemas-upnp-org:device:ZonePlayer:1</deviceType>
<friendlyName>192.168.1.180 - Sonos PLAY:1</friendlyName>
<manufacturer>Sonos, Inc.</manufacturer>
<manufacturerURL>http://www.sonos.com</manufacturerURL>
<modelNumber>S12</modelNumber>
<modelDescription>Sonos PLAY:1</modelDescription>
<modelName>Sonos PLAY:1</modelName>
<modelURL>http://www.sonos.com/products/zoneplayers/S12</modelURL>
<softwareVersion>43.3-53160</softwareVersion>
<hardwareVersion>1.20.1.6-2</hardwareVersion>
<serialNum>XXXXXXXXXXXXXXXXX</serialNum>
<UDN>uuid:RINCON_XXXXXXXXXXXXXXXXX</UDN>
<iconList>
<icon>
<id>0</id>
<mimetype>image/png</mimetype>
<width>48</width>
<height>48</height>
<depth>24</depth>
<url>/img/icon-S12.png</url>
</icon>
</iconList>
<minCompatibleVersion>42.0-00000</minCompatibleVersion>
<legacyCompatibleVersion>36.0-00000</legacyCompatibleVersion>
<apiVersion>1.5.2</apiVersion>
<minApiVersion>1.1.0</minApiVersion>
<displayVersion>8.6</displayVersion>
<extraVersion/>
<roomName>Office</roomName>
<displayName>PLAY:1</displayName>
<zoneType>14</zoneType>
<feature1>0x00000000</feature1>
<feature2>0x00403332</feature2>
<feature3>0x0001000e</feature3>
<seriesid>A200</seriesid>
<variant>2</variant>
<internalSpeakerSize>5</internalSpeakerSize>
<bassExtension>75.000</bassExtension>
<satGainOffset>6.000</satGainOffset>
<memory>256</memory>
<flash>256</flash>
<ampOnTime>10</ampOnTime>
<retailMode>0</retailMode>
<serviceList>
<service>
<serviceType>urn:schemas-upnp-org:service:AlarmClock:1</serviceType>
<serviceId>urn:upnp-org:serviceId:AlarmClock</serviceId>
<controlURL>/AlarmClock/Control</controlURL>
<eventSubURL>/AlarmClock/Event</eventSubURL>
<SCPDURL>/xml/AlarmClock1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:MusicServices:1</serviceType>
<serviceId>urn:upnp-org:serviceId:MusicServices</serviceId>
<controlURL>/MusicServices/Control</controlURL>
<eventSubURL>/MusicServices/Event</eventSubURL>
<SCPDURL>/xml/MusicServices1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:DeviceProperties:1</serviceType>
<serviceId>urn:upnp-org:serviceId:DeviceProperties</serviceId>
<controlURL>/DeviceProperties/Control</controlURL>
<eventSubURL>/DeviceProperties/Event</eventSubURL>
<SCPDURL>/xml/DeviceProperties1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:SystemProperties:1</serviceType>
<serviceId>urn:upnp-org:serviceId:SystemProperties</serviceId>
<controlURL>/SystemProperties/Control</controlURL>
<eventSubURL>/SystemProperties/Event</eventSubURL>
<SCPDURL>/xml/SystemProperties1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:ZoneGroupTopology:1</serviceType>
<serviceId>urn:upnp-org:serviceId:ZoneGroupTopology</serviceId>
<controlURL>/ZoneGroupTopology/Control</controlURL>
<eventSubURL>/ZoneGroupTopology/Event</eventSubURL>
<SCPDURL>/xml/ZoneGroupTopology1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:GroupManagement:1</serviceType>
<serviceId>urn:upnp-org:serviceId:GroupManagement</serviceId>
<controlURL>/GroupManagement/Control</controlURL>
<eventSubURL>/GroupManagement/Event</eventSubURL>
<SCPDURL>/xml/GroupManagement1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-tencent-com:service:QPlay:1</serviceType>
<serviceId>urn:tencent-com:serviceId:QPlay</serviceId>
<controlURL>/QPlay/Control</controlURL>
<eventSubURL>/QPlay/Event</eventSubURL>
<SCPDURL>/xml/QPlay1.xml</SCPDURL>
</service>
</serviceList>
<deviceList>
<device>
<deviceType>urn:schemas-upnp-org:device:MediaServer:1</deviceType>
<friendlyName>192.168.1.180 - Sonos PLAY:1 Media Server</friendlyName>
<manufacturer>Sonos, Inc.</manufacturer>
<manufacturerURL>http://www.sonos.com</manufacturerURL>
<modelNumber>S12</modelNumber>
<modelDescription>Sonos PLAY:1 Media Server</modelDescription>
<modelName>Sonos PLAY:1</modelName>
<modelURL>http://www.sonos.com/products/zoneplayers/S12</modelURL>
<UDN>uuid:RINCON_XXXXXXXXXXXXXXXXX_MS</UDN>
<serviceList>
<service>
<serviceType>urn:schemas-upnp-org:service:ContentDirectory:1</serviceType>
<serviceId>urn:upnp-org:serviceId:ContentDirectory</serviceId>
<controlURL>/MediaServer/ContentDirectory/Control</controlURL>
<eventSubURL>/MediaServer/ContentDirectory/Event</eventSubURL>
<SCPDURL>/xml/ContentDirectory1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:ConnectionManager:1</serviceType>
<serviceId>urn:upnp-org:serviceId:ConnectionManager</serviceId>
<controlURL>/MediaServer/ConnectionManager/Control</controlURL>
<eventSubURL>/MediaServer/ConnectionManager/Event</eventSubURL>
<SCPDURL>/xml/ConnectionManager1.xml</SCPDURL>
</service>
</serviceList>
</device>
<device>
<deviceType>urn:schemas-upnp-org:device:MediaRenderer:1</deviceType>
<friendlyName>Office - Sonos PLAY:1 Media Renderer</friendlyName>
<manufacturer>Sonos, Inc.</manufacturer>
<manufacturerURL>http://www.sonos.com</manufacturerURL>
<modelNumber>S12</modelNumber>
<modelDescription>Sonos PLAY:1 Media Renderer</modelDescription>
<modelName>Sonos PLAY:1</modelName>
<modelURL>http://www.sonos.com/products/zoneplayers/S12</modelURL>
<UDN>uuid:RINCON_XXXXXXXXXXXXXXXXX_MR</UDN>
<serviceList>
<service>
<serviceType>urn:schemas-upnp-org:service:RenderingControl:1</serviceType>
<serviceId>urn:upnp-org:serviceId:RenderingControl</serviceId>
<controlURL>/MediaRenderer/RenderingControl/Control</controlURL>
<eventSubURL>/MediaRenderer/RenderingControl/Event</eventSubURL>
<SCPDURL>/xml/RenderingControl1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:ConnectionManager:1</serviceType>
<serviceId>urn:upnp-org:serviceId:ConnectionManager</serviceId>
<controlURL>/MediaRenderer/ConnectionManager/Control</controlURL>
<eventSubURL>/MediaRenderer/ConnectionManager/Event</eventSubURL>
<SCPDURL>/xml/ConnectionManager1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:AVTransport:1</serviceType>
<serviceId>urn:upnp-org:serviceId:AVTransport</serviceId>
<controlURL>/MediaRenderer/AVTransport/Control</controlURL>
<eventSubURL>/MediaRenderer/AVTransport/Event</eventSubURL>
<SCPDURL>/xml/AVTransport1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-sonos-com:service:Queue:1</serviceType>
<serviceId>urn:sonos-com:serviceId:Queue</serviceId>
<controlURL>/MediaRenderer/Queue/Control</controlURL>
<eventSubURL>/MediaRenderer/Queue/Event</eventSubURL>
<SCPDURL>/xml/Queue1.xml</SCPDURL>
</service>
<service>
<serviceType>
urn:schemas-upnp-org:service:GroupRenderingControl:1
</serviceType>
<serviceId>urn:upnp-org:serviceId:GroupRenderingControl</serviceId>
<controlURL>/MediaRenderer/GroupRenderingControl/Control</controlURL>
<eventSubURL>/MediaRenderer/GroupRenderingControl/Event</eventSubURL>
<SCPDURL>/xml/GroupRenderingControl1.xml</SCPDURL>
</service>
<service>
<serviceType>urn:schemas-upnp-org:service:VirtualLineIn:1</serviceType>
<serviceId>urn:upnp-org:serviceId:VirtualLineIn</serviceId>
<controlURL>/MediaRenderer/VirtualLineIn/Control</controlURL>
<eventSubURL>/MediaRenderer/VirtualLineIn/Event</eventSubURL>
<SCPDURL>/xml/VirtualLineIn1.xml</SCPDURL>
</service>
</serviceList>
<X_Rhapsody-Extension xmlns="http://www.real.com/rhapsody/xmlns/upnp-1-0">
<deviceID>
urn:rhapsody-real-com:device-id-1-0:sonos_1:RINCON_XXXXXXXXXXXXXXXXX
</deviceID>
<deviceCapabilities>
<interactionPattern type="real-rhapsody-upnp-1-0"/>
</deviceCapabilities>
</X_Rhapsody-Extension>
<qq:X_QPlay_SoftwareCapability xmlns:qq="http://www.tencent.com">QPlay:2</qq:X_QPlay_SoftwareCapability>
<iconList>
<icon>
<mimetype>image/png</mimetype>
<width>48</width>
<height>48</height>
<depth>24</depth>
<url>/img/icon-S12.png</url>
</icon>
</iconList>
</device>
</deviceList>
</device>
</root>
```

File: mobile/src/main/java/net/mm2d/android/upnp/avt/MrControlPoint.java [hunk 1]
```diff
@@ -86,13 +86,20 @@ protected MediaRenderer createMediaRenderer(@NonNull final Device device) {
...
-                || device.findServiceById(Avt.AVT_SERVICE_ID) == null) {
+        Device realDevice = device;
+        for (Device subDevice : device.getDeviceList()) {
+            if (subDevice.getDeviceType().startsWith(Avt.MR_DEVICE_TYPE)) {
+                realDevice = subDevice;
+                break;
+            }
+        }
+        if (!realDevice.getDeviceType().startsWith(Avt.MR_DEVICE_TYPE)
+                || realDevice.findServiceById(Avt.AVT_SERVICE_ID) == null) {
```
Comment (ohmae) [inline mobile/src/main/java/net/mm2d/android/upnp/avt/MrControlPoint.java:97]: How about separating the check part of Device into another method?
```java

    private void discoverDevice(@NonNull final Device device) {
        if (device.getDeviceType().startsWith(Avt.MR_DEVICE_TYPE)
                && device.findServiceById(Avt.AVT_SERVICE_ID) != null) {
            discoverMrDevice(device);
            return;
        }
        for (final Device subDevice : device.getDeviceList()) {
            if (subDevice.getDeviceType().startsWith(Avt.MR_DEVICE_TYPE)
                    && subDevice.findServiceById(Avt.AVT_SERVICE_ID) != null) {
                discoverMrDevice(subDevice);
                break;
            }
        }
    }
```

File: mobile/src/main/java/net/mm2d/android/upnp/avt/MrControlPoint.java [hunk 2]
```diff
@@ -86,13 +86,20 @@ protected MediaRenderer createMediaRenderer(@NonNull final Device device) {
     }
 
     private void discoverDevice(@NonNull final Device device) {
-        if (!device.getDeviceType().startsWith(Avt.MR_DEVICE_TYPE)
-                || device.findServiceById(Avt.AVT_SERVICE_ID) == null) {
+        Device realDevice = device;
+        for (Device subDevice : device.getDeviceList()) {
```
Comment (ohmae) [inline mobile/src/main/java/net/mm2d/android/upnp/avt/MrControlPoint.java:90]: Is it not necessary to check recursively about descendant devices?

**Review by ohmae — COMMENTED** [review]
Please check my comments

**Comment by ohmae** [comment]
Thank you for teaching XML. I am saved.

**Comment by Tolriq** [comment]
Not sure to understand both questions :)

I did only find that missing part when testing this app before integrating mmupnp in my own app. So thought you might want the info and possible fix about some special devices.

I'm can't really pass too much time tweaking the code, you are free to use it as you wish since I also provided the device root upnp xml so you can reproduce in tests.

**Comment by ohmae** [comment]
I understand your intention.
I will merge your RP. and I will use it as I like. :smile: 
Thank you very much.

**Pull request merged**